video/clip.go

```diff
@@ -0,0 +1,72 @@
+package video
+
+import (
+	"fmt"
+	"github.com/grafov/m3u8"
+	"github.com/livepeer/catalyst-api/log"
+)
+
+// Checks if a given segment already exists in a slice of MediaSegments. Mainly used to ensure
+// we don't add duplicate segments when the playhead start/stop time falls in the same segment.
+func segmentExists(relSegs []*m3u8.MediaSegment, seg *m3u8.MediaSegment) bool {
+	for _, s := range relSegs {
+		if s.SeqId == seg.SeqId {
+			return true
+		}
+	}
+	return false
+}
+
+// Finds the segment in an HLS manifest that contains the timestamp (aka playhead) specified.
+func getRelevantSegment(allSegments []*m3u8.MediaSegment, relevantSegments *[]*m3u8.MediaSegment, playHeadTime float64, startingIdx uint64) (segNum uint64, err error) {
+	playHeadDiff := 0.0
+	segNum = 0
+	for _, segment := range allSegments[startingIdx:] {
+		// Break if we reach the end of the MediaSegment slice that contains all segments in the manifest
+		if segment == nil {
+			break
+		}
+		// Skip any segments where duration is 0s
+		if segment.Duration <= 0.0 {
+			return 0, fmt.Errorf("error clipping: found 0s duration segments")
+		}
+		// Check if the playhead is within the current segment and skip to
+		// the next segment if it's not. Also update the play head by referencing
+		// the starting time of the next segment.
+		playHeadDiff = playHeadTime - segment.Duration
+		if playHeadDiff > 0.0 {
+			playHeadTime = playHeadDiff
+			continue
+		}
+		// Skip adding to slice of relevantSegments if the segment already exists
+		// (e.g. start and end times fall within same segment index)
```

Comment:
if this is the only case where this is possible maybe we could just increment `startingIdx` by one and exit if we've gone past the `endTime`. Or check that the current index is equal to `startingIdx`. Then we could avoid looping through `relevantSegments`

Reply:
Updated the logic to simplify all of this further.